### ISSUE TITLE
Unit testing HTTP Callable functions

### ISSUE THREAD

**Issue by ahaverty** [firebase/firebase-functions-test#9]
I'm currently moving a lot of our functions to the new HTTP callable functions, and noticed that the [docs note](https://firebase.google.com/docs/functions/unit-testing#testing_http_functions): 

> Note: Testing of HTTP Callable Functions is not yet supported.

Is this something that might be added to the library soon or is there some long-term limitation?

**Comment by laurenzlong** [comment]
We're working on it!

**Comment by ahaverty** [comment]
@laurenzlong Sorry for github issue spam, but adding unit testing of https.onCall is now at the top of our companies firebase wish-list 🙏 We'd appreciate hearing about any updates 😄

**Comment by g-viet** [comment]
we could take our all logic part inside of the callable function into another function.
Testing them first. Then wait any updates 😄

**Comment by laurenzlong** [comment]
Sorry for the delay, and thanks for the patience everyone! We've been busy working on launching regional support, Node 8 support, and configuration of timeouts and memory allocation, now that those are out the door, this is next up on our roadmap.

**Comment by joelgetaction** [comment]
I really need this since most of my functions are http callable.  In fact I'd be happy to help with this if you can assign me some tasks or hints ... :-)

Any updates?

**Comment by laurenzlong** [comment]
Thanks for the patience everyone! Our team is stretched thin right now so we haven't had a chance to implement this yet. However, you can make use of the `run` method offered by the firebase-functions SDK, which exists on all cloud functions except HTTPS onRequest functions (see[ code](url) for callable functions)

For example, let's say you have a callable function called `myCallableFunction`

```
const myFunctions = require('../index');
const expect = require('chai').expect;

const data = {foo: 'bar'};
const callableContext = {auth: {uid: 'abc'}}; // you can also add "rawRequest" or "instanceIdToken" fields

expect(myFunctions.myCallableFunction.run(data, callableContext))
  .to.eventually.deep.equal({expected: "result"});

```

**Comment by sipersso** [comment]
@laurenzlong Do I understand it correctly that there is no way to test HTTPS callable?

**Comment by gustavopch** [comment]
@sipersso The `firebase-functions-test` package doesn't currently provide tools for testing Callable Functions, but you can test them anyways as @laurenzlong shows in the snippet right before your comment.

**Comment by sipersso** [comment]
@gustavopch yes, I tried it and myFunctions.myCallableFunction.run is not "not a function", hence my question. 

Since @laurenzlong wrote that "which exists on all cloud functions except HTTPS onRequest functions", I figured that HTTPS was the key here and not onRequest.

**Comment by ericvera** [comment]
Was this already released? Documented? I tried for more than an hour to get it to work v0.1.6, but couldn't. I've been using the .run( trick, but that seems to skip over things like mocking config values. :(

**Comment by AndrsDev** [comment]
Is there any documentation of this new feature about how to test callable functions? I've found none in the [documentation](https://firebase.google.com/docs/functions/unit-testing#testing_http_functions).

**Comment by tiendn** [comment]
@AndrsDev  Seem like, Callable not yet ready. 
https://firebase.google.com/docs/functions/unit-testing#testing_http_functions

**Comment by hireplacer** [comment]
@tiendn , I think this is how you test callable functions: https://github.com/firebase/firebase-functions-test/blob/c77aa92d345b8e4fb5ad98534989eb8dcf7d9bc4/spec/providers/https.spec.ts

**Comment by tiendn** [comment]
Thanks @hireplacer

**Comment by leomendoza123** [comment]
This is great but why almost one year later the official docs still says that this is not supported? @abeisgoat 

https://firebase.google.com/docs/functions/unit-testing#testing_http_functions

**Comment by r3khaan** [comment]
@abeisgoat Hey, any updates on why this is not in the official documentation?

**Comment by SpiegelSoft** [comment]
If I include the `options` argument, I get this error message when I try to test the callable function:

Value for argument "data" is not a valid Firestore document. Cannot use "undefined" as a Firestore value (found in field "quota").

**Comment by lengocgiang** [comment]
from @laurenzlong anwser, I added more detail with chai-promise
```javascript
const chai = require('chai') ;
const chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
const expect = require('chai').expect;
const myFunctions = require('../index');
describe('test http callable functions', () => {
    it('test your http callable function', async () => {
        const data = {
            "id":1
        }
        const callableContext = {auth:{uid:'abc'}};
        const expectedResponse = {
            error:false,
            body:"ok",
        };
        await expect(myFunctions.{your_function_name}.run(data,context)).to.eventually.deep.equal({expectedResponse});
    })
})
```
